stan/math/rev/mat/fun/multiply.hpp

```diff
@@ -1,56 +1,560 @@
+          : vari(0.0),
+          A_cols_(A.cols()),
+          Ad_(ChainableStack::memalloc_.alloc_array<double>(A_cols_)),
+          Bd_(ChainableStack::memalloc_.alloc_array<double>(A_cols_)),
+          variRefA_(ChainableStack::memalloc_.alloc_array<vari*>(A_cols_)) {
+            using Eigen::Map;
+            using Eigen::VectorXd;
+            using Eigen::RowVectorXd;
+            for (size_type i = 0; i < A_cols_; ++i) {
+              variRefA_[i] = A.coeffRef(i).vi_;
+              Ad_[i] = A.coeffRef(i).val();
+            }
+            for (size_type i = 0; i < A_cols_; ++i)
+              Bd_[i] = B.coeffRef(i);
+            double AB
+              = Map<RowVectorXd>(Ad_, 1, A_cols_)
+              * Map<VectorXd>(Bd_, A_cols_, 1);
+            variRefAB_ = new vari(AB, false);
+          }
+
+        virtual void chain() {
+          using Eigen::Map;
+          using Eigen::VectorXd;
+          using Eigen::RowVectorXd;
+          double adjAB;
+          RowVectorXd adjA(A_cols_);
+
+          adjAB = variRefAB_->adj_;
+          adjA = adjAB
+            * Map<VectorXd>(Bd_, A_cols_, 1).transpose();
+          for (size_type i = 0; i < A_cols_; ++i)
+            variRefA_[i]->adj_ += adjA(i);
+        }
+    };
+
+    /**
      * Return the product of two scalars.
-     * @param[in] v First scalar.
-     * @param[in] c Specified scalar.
-     * @return Product of scalars.
+     * @tparam T1 scalar type of v
+     * @tparam T2 scalar type of c
+     * @param[in] v First scalar
+     * @param[in] c Specified scalar
+     * @return Product of scalars
      */
     template <typename T1, typename T2>
     inline typename
     boost::enable_if_c<
-      (boost::is_scalar<T1>::value || boost::is_same<T1, var>::value)
-      && (boost::is_scalar<T2>::value || boost::is_same<T2, var>::value),
-      typename boost::math::tools::promote_args<T1, T2>::type>::type
-             multiply(const T1& v, const T2& c) {
-      return v * c;
-    }
+    (boost::is_scalar<T1>::value || boost::is_same<T1, var>::value)
+    && (boost::is_scalar<T2>::value || boost::is_same<T2, var>::value),
+    typename boost::math::tools::promote_args<T1, T2>::type>::type
+      multiply(const T1& v, const T2& c) {
+        return v * c;
+      }
 
     /**
      * Return the product of scalar and matrix.
-     * @param[in] c Specified scalar.
-     * @param[in] m Matrix.
-     * @return Product of scalar and matrix.
+     * @tparam T1 scalar type v
+     * @tparam T2 scalar type matrix m
+     * @tparam R2 Rows matrix m
+     * @tparam C2 Columns matrix m
+     * @param[in] c Specified scalar
+     * @param[in] m Matrix
+     * @return Product of scalar and matrix
      */
     template<typename T1, typename T2, int R2, int C2>
     inline Eigen::Matrix<var, R2, C2>
     multiply(const T1& c, const Eigen::Matrix<T2, R2, C2>& m) {
-      // FIXME:  pull out to eliminate overpromotion of one side
```

Comment:
If we want to follow Google style guide, it'd be

TODO(trangucci): blah blah blah